Initial commit for swift-ogg

langleyd opened the pull request:
This PR adds support for converting between `opus/ogg` and `mpeg4aac/m4a` files.
Most files are opus' own header files.
The main files of interest are `OGGEncoder` and `OGGDecoder` which were adapted from [watson-developer-cloud/swift-sdk](https://github.com/watson-developer-cloud/swift-sdk). Within that library they were specifcally targeted at the a speed -to-text use case and converted directly to WAV format rather than raw PCM(with no WAV container).
This encoder/decoder within this convert directly to linear PCM which is then used in [OGGConverted](https://github.com/vector-im/swift-ogg/compare/langleyd/initial?expand=1#diff-23e6e2d2e94822cd92e290563b1ba0e10ca22fd6a984fe0edf1c6eb51b5fe3a7R26) along with apples native `AVFAudio` and `AVFoundation` libraries to convert between PCM and `mpeg4aac/m4a`

Round-trip conversion tests are added in [ConverterTests](https://github.com/vector-im/swift-ogg/compare/langleyd/initial?expand=1#diff-4bd0796b9ac17a3a924acfbb31aa6da07c5c1c3f767fe8d1220beaf9ad6a5582R21) using a an audio recording of 2 minutes(at the time of writing the max supported for element voice messages).

stefanceriu commented on README.md:
```diff
@@ -1,6 +1,22 @@
-A Swift library for converting to and from opus/ogg format
+This library provides a very simple api to convert between `opus/ogg` and `MPEG4AAC/m4a` files.
+It uses opus's `libopus` `libogg` to convert between `opus/ogg` and linear PCM audio data and native iOS apis to convert between PCM and `MPEG4AAC/m4a`(but could be very easily adapted to use any other codec/container supported by iOS).
+Apples recording and playback apis form `AVFoundation` can then be used quite simply with a codec/container pair that iOS natively supports(`MPEG4AAC/m4a` in this case).
```
```suggestion
Apple's recording and playback APIs from `AVFoundation` can then be used quite simply with a codec/container pair that iOS natively supports (`MPEG4AAC/m4a` in this case).
```

stefanceriu commented on README.md:
```diff
@@ -1,6 +1,22 @@
 # swift-ogg
-A Swift library for converting to and from opus/ogg format
+This library provides a very simple api to convert between `opus/ogg` and `MPEG4AAC/m4a` files.
```
```suggestion
This library provides a very simple API to convert between `opus/ogg` and `MPEG4AAC/m4a` files.
```

stefanceriu commented on Sources/SwiftOgg/OGGDecoder.swift:
```diff
@@ -0,0 +1,281 @@
+ **/
+
+import Foundation
+#if canImport(Clibogg)
```
Is there any point in checking the imports if everything else would fail when not available? 🤔

stefanceriu commented on Sources/SwiftOgg/OGGConverter.swift:
```diff
@@ -0,0 +1,85 @@
+import AVFAudio
+import AVFoundation
+
+public enum OGGConverterError: Error {
```
It would be nice if this would only expose its own errors instead of forwarding ones from undearneath e.g. `OpusError.invalidState`, `OggError.internalError`, even if only wrapping them.

aringenbach replied:
SGTM too (there is a lot of them though, probably just wrapping them is good enough ? - I suppose these won't be that relevant to clients using this)

stefanceriu reviewed (approved):
I'm not going to pretend I understand everything this does but I trust the tests to do their job. It looks really good to me, well done! 👏

aringenbach commented on Sources/SupportingFiles/Dependencies/Copustools/include/ogg.h:
```diff
@@ -0,0 +1,209 @@
+ *                                                                  *
+ * THIS FILE IS PART OF THE OggVorbis SOFTWARE CODEC SOURCE CODE.   *
+ * USE, DISTRIBUTION AND REPRODUCTION OF THIS LIBRARY SOURCE IS     *
+ * GOVERNED BY A BSD-STYLE SOURCE LICENSE INCLUDED WITH THIS SOURCE *
+ * IN 'COPYING'. PLEASE READ THESE TERMS BEFORE DISTRIBUTING.       *
```
Looks like this is only on a few headers but perhaps we should include a copy of this license ?

langleyd replied:
Yea I reference this licence in the readme and will include in the element-ios [list of third party licences](https://github.com/vector-im/element-ios/blob/develop/Riot/Assets/third_party_licenses.html). That what you are referring to or something else additionally?

aringenbach commented on Sources/SwiftOgg/OGGEncoder.swift:
```diff
@@ -0,0 +1,469 @@
+            try encode(pcm: bytes, count: pcm.count)
+        }
+
+        // try encode(pcm: pcm.bytes, bytes: pcm.length)
```
is this something to remove ?

langleyd commented:
@stefanceriu @aringenbach Addressed comments and also moved to using https://github.com/ybrid/opus-swift
https://github.com/ybrid/ogg-swift which package the opus libraries as xcframeworks and support spm, which will simplify build/ci.
